packages/sandcastle/src/Gallery.tsx

```diff
@@ -15,6 +16,140 @@ export type GalleryItem = {
   isNew: boolean;
 };
 
+export function GallerySearch({
+  setSearchResults,
+  setTag,
+}: {
+  setSearchResults: (newSearchResults: PagefindSearchFragment[] | null) => void;
+  setTag: (newTag: string | "All") => void;
+}) {
+  const pagefind = useRef<Pagefind>(null);
+  const isFirstLoad = useRef(true);
+  const [pagefindLoaded, setPagefindLoaded] = useState(false);
+  const [availableFilters, setAvailableFilters] =
+    useState<PagefindFilterCounts>({ tags: {} });
+
+  const [searchTerm, setSearchTerm] = useState("");
+  const [currentTag, setCurrentTag] = useState<string | "All">("All");
+
+  const requestId = useRef<number>(0);
+
+  useEffect(() => {
+    async function doSearch() {
+      if (!pagefind.current) {
+        return;
+      }
+      const searchString = searchTerm;
+      if (!searchString || searchString === "") {
+        setSearchResults(null);
+        return;
+      }
+      const thisRequestId = ++requestId.current;
+
+      const search = await pagefind.current.search(searchString, {
+        filters: {
+          tags: currentTag === "All" ? undefined : currentTag,
+        },
+      });
+
+      const resultDataPromises = await Promise.allSettled(
+        search.results.map((result) => {
+          return result.data();
+        }),
+      );
+      const resultData = resultDataPromises
+        .map((result) => {
+          return result.status === "fulfilled" ? result.value : undefined;
+        })
+        .filter((result) => !!result);
+
+      if (thisRequestId === requestId.current) {
+        setSearchResults(resultData);
+      }
+    }
+
+    doSearch();
+  }, [searchTerm, currentTag, setSearchResults]);
+
+  useEffect(() => {
+    async function loadPageFind() {
+      const pagefindImport: Pagefind = await import(
+        // vite doesn't like the dynamic import
+        /* @vite-ignore */
+        `${GALLERY_BASE}/pagefind/pagefind.js`
+      );
+
+      pagefindImport.init();
+      await pagefindImport.options({
+        baseUrl: getBaseUrl(),
+      });
+      pagefind.current = pagefindImport;
+
+      const filters = await pagefind.current.filters();
+      setAvailableFilters(filters);
+
+      setPagefindLoaded(true);
+    }
+    if (isFirstLoad.current) {
+      // This prevents double activation when using StrictMode
+      isFirstLoad.current = false;
+      loadPageFind();
+    }
+  }, []);
+
+  const debounceTimeout = useRef<NodeJS.Timeout>(null);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  return (
+    <>
+      {!!searchTerm && (
+        <Button
+          onClick={() => {
+            setSearchTerm("");
+            if (inputRef.current) {
+              inputRef.current.value = "";
+            }
+          }}
+        >
+          Clear
+        </Button>
```

Comment:
<img width="243" height="47" alt="Image" src="https://github.com/user-attachments/assets/6a83d911-34d6-49a2-ad6f-181a8131a106" /> 

Instead of the "Clear" button (which is a bit non-standard AFAIK), can we implement a more minimal cancel icon button?

i.e., something like the following (It's fine to stick with "clear" as a label IMO) .

<img width="374" height="62" alt="Image" src="https://github.com/user-attachments/assets/3ab6d542-46aa-4d39-80a9-6c479d74c884" />

<img width="281" height="95" alt="Image" src="https://github.com/user-attachments/assets/42bd4a7c-560b-40fc-af22-3778529508df" />

Reply:
I also noticed both of these examples use the search icon as well. I would suggest we add it. I think that would help guide users looking for search capabilities.